### PR TITLE
Added password reset functionality

### LINKED ISSUE #1600: Add ability to reset the user password.
### My actions before raising this issue
- [x] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [x] Searched [past issues](/issues)

Currently there is no ability to reset password.

### Expected Behaviour
Should be possible reset the password.

### Current Behaviour
NA

### Steps to Reproduce (for bugs)
NA

### Context
NA

### Your Environment
NA

### Next steps
You may [join our Gitter](https://gitter.im/opencv-cvat/public) channel for community support.

### PR DESCRIPTION
**Pull request by azhavoro**
resolve #1600

Added password reset functionality

### How has this been tested?

manual testing

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by bsekachev** [comment]
@azhavoro Please, resolve some conflicts

File: cvat-ui/src/components/reset-password-page/reset-password-page.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                        <Col>
+                            <Text strong>
+                                Already have an account?
+                                <Link to='/auth/login'> Login </Link>
+                            </Text>
+                        </Col>
+                    </Row>
+                </Col>
+            </Row>
+        </>
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-page/reset-password-page.tsx:74]: Do not need <> </> here

File: cvat-ui/src/components/reset-password-page/reset-password-page.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    <ResetPasswordForm
+                        fetching={fetching}
+                        onSubmit={(resetPasswordData: ResetPasswordData): void => {
+                            onResetPassword(resetPasswordData.email);
+                        }}
+                    />
+                    <Row type='flex' justify='start' align='top'>
+                        <Col>
+                            <Text strong>
+                                Already have an account?
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-page/reset-password-page.tsx:67]: I suppose this text isn't expected here. Something like ``Go to login page`` would be better

File: cvat-ui/src/components/reset-password-page/reset-password-form.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    />,
+                )}
+            </Form.Item>
+        );
+    }
+
+    public render(): JSX.Element {
+        const { fetching } = this.props;
+        return (
+            <Form onSubmit={this.handleSubmit} className='reset-password-form'>
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-page/reset-password-form.tsx:62]: Let's add prefix ``cvat`` to all used css classes

File: cvat-ui/src/components/reset-password-page/reset-password-form.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                {this.renderEmailField()}
+
+                <Form.Item>
+                    <Button
+                        type='primary'
+                        loading={fetching}
+                        disabled={fetching}
+                        htmlType='submit'
+                        className='reset-password-form-button'
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-page/reset-password-form.tsx:71]: The same comment about css classes

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                        <Col>
+                            <Text strong>
+                                Already have an account?
+                                <Link to='/auth/login'> Login </Link>
+                            </Text>
+                        </Col>
+                    </Row>
+                </Col>
+            </Row>
+        </>
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx:86]: React fragment <></> isn't needed here

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        const uid = params.get('uid');
+        const token = params.get('token');
+
+        form.validateFields((error, values): void => {
+            if (!error) {
+                const validatedFields = {
+                    ...values,
+                    uid,
+                    token,
+                    confirmations: [],
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:78]: Why do we need this empty array?

File: cvat-ui/src/reducers/notifications-reducer.ts [hunk 7]
```diff
@@ -190,6 +194,61 @@ export default function (state = defaultState, action: AnyAction): Notifications
...
         }
+        case AuthActionTypes.RESET_PASSWORD_SUCCESS: {
+            return {
+                ...state,
+                messages: {
+                    ...state.messages,
+                    auth: {
+                        ...state.messages.auth,
+                        resetPasswordDone: `If your email address exists in our database,
+                            you will receive a password recovery link at your email address in a few minutes.`,
```
[the remark this answers is not in the document]
Comment (azhavoro) [reply cvat-ui/src/reducers/notifications-reducer.ts:205]: This is a security leak. I think it's best to leave the default behavior.

File: cvat-core/src/api.js [hunk 8]
```diff
@@ -206,6 +209,38 @@ function build() {
...
+                * @async
+                * @memberof module:API.cvat.server
+                * @param {string} email A email address for the account
+                * @throws {module:API.cvat.exceptions.PluginError}
+                * @throws {module:API.cvat.exceptions.ServerError}
+            */
+            async resetPassword(email) {
+                const result = await PluginRegistry
+                    .apiWrapper(cvat.server.resetPassword, email);
+                return result;
```
[the remark this answers is not in the document]
Comment (azhavoro) [reply cvat-core/src/api.js:224]: sure

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx (hunk 6 above)
Comment (azhavoro) [reply cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:78]: My bad, ctrl+c, ctrl+v....

File: cvat-ui/src/reducers/notifications-reducer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/reducers/notifications-reducer.ts:205]: Brutforce to get to know all registered emails? What about Debouncing policies?

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                                resetPasswordConfirmData.newPassword2,
+                                resetPasswordConfirmData.uid,
+                                resetPasswordConfirmData.token,
+                            );
+                        }}
+                    />
+                    <Row type='flex' justify='start' align='top'>
+                        <Col>
+                            <Text strong>
+                                Already have an account?
```
Comment (bsekachev) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx:79]: I suppose this text isn't expected here. Something like Go to login page would be better

File: cvat-ui/src/reducers/notifications-reducer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (azhavoro) [reply cvat-ui/src/reducers/notifications-reducer.ts:205]: One of the options, but information about whether the address is registered or not is a leak. We have already enabled throttling policies for unauthenticated requests like this.

**Comment by bsekachev** [comment]
LGTM, @ActiveChooN, could you please also take a look to this PR?

File: cvat-ui/src/reducers/notifications-reducer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/reducers/notifications-reducer.ts]: We discussed it on the meeting

**Review by bsekachev — APPROVED** [review]

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+type ResetPasswordConfirmFormProps = {
+    fetching: boolean;
+    onSubmit(resetPasswordConfirmData: ResetPasswordConfirmData): void;
+} & FormComponentProps & RouteComponentProps;
+
+class ResetPasswordConfirmFormComponent extends React.PureComponent<ResetPasswordConfirmFormProps> {
+    private validateConfirmation = (_: any, value: string, callback: Function): void => {
+        const { form } = this.props;
+        if (value && value !== form.getFieldValue('newPassword1')) {
+            callback('Two passwords that you enter is inconsistent!');
```
Comment (ActiveChooN) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:30]: `do not match` would be better, i guess.

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                            resetPasswordConfirmData.newPassword2,
+                            resetPasswordConfirmData.uid,
+                            resetPasswordConfirmData.token,
+                        );
+                    }}
+                />
+                <Row type='flex' justify='start' align='top'>
+                    <Col>
+                        <Text strong>
+                            Already have an account?
```
Comment (ActiveChooN) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx:78]: Looks like question for the register page. Why do we need it here?

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx (hunk 10 above)
Comment (nmanovic) [reply cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:30]: +1 `Passwords do not match`

Comment (nmanovic) [inline cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-form.tsx:30]:
```suggestion
            callback('Passwords do not match!');
```

File: cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx (hunk 11 above)
Comment (azhavoro) [reply cvat-ui/src/components/reset-password-confirm-page/reset-password-confirm-page.tsx]: Ok, lets remove it

**Review by ActiveChooN — APPROVED** [review]

**Review by bsekachev — APPROVED** [review]

**Pull request merged**